community/detectors/roxy_wi_cve_2022_31137/src/main/java/com/google/tsunami/plugins/detectors/cves/cve202231137/Cve202231137Detector.java

```diff
@@ -0,0 +1,170 @@
+    this.payloadGenerator = checkNotNull(payloadGenerator);
+  }
+
+  @Override
+  public DetectionReportList detect(
+      TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+    logger.atInfo().log("Cve202231137Detector starts detecting.");
+
+    return DetectionReportList.newBuilder()
+        .addAllDetectionReports(
+            matchedServices.stream()
+                .filter(NetworkServiceUtils::isWebService)
+                .filter(this::isRoxyWiWebService)
+                .filter(this::isServiceVulnerable)
+                .map(networkService -> buildDetectionReport(targetInfo, networkService))
+                .collect(toImmutableList()))
+        .build();
+  }
+
+  private boolean isRoxyWiWebService(NetworkService networkService) {
+    // note that this fingerprint phase is only for vulnerable versions,
+    // because newer Roxy-Wi versions have different endpoints for login
+    String targetUrl =
+        NetworkServiceUtils.buildWebApplicationRootUrl(networkService) + "app/login.py";
+    try {
+      Optional<String> response =
+          httpClient.send(get(targetUrl).withEmptyHeaders().build(), networkService).bodyString();
+      return response.isPresent() && response.get().contains("<title>Login page - Roxy-WI</title>");
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Request to target %s failed", networkService);
+    }
+    return false;
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
+    PayloadGeneratorConfig config =
+        PayloadGeneratorConfig.newBuilder()
+            .setVulnerabilityType(PayloadGeneratorConfig.VulnerabilityType.REFLECTIVE_RCE)
+            .setInterpretationEnvironment(
+                PayloadGeneratorConfig.InterpretationEnvironment.LINUX_SHELL)
+            .setExecutionEnvironment(
+                PayloadGeneratorConfig.ExecutionEnvironment.EXEC_INTERPRETATION_ENVIRONMENT)
+            .build();
+
+    Payload payload = payloadGenerator.generate(config);
+    String cmd = payload.getPayload();
+
+    HttpResponse response = null;
+    String targetVulnerabilityUrl =
+        NetworkServiceUtils.buildWebApplicationRootUrl(networkService) + VULNERABLE_REQUEST_PATH;
+    try {
+      response =
+          httpClient.send(
+              post(targetVulnerabilityUrl)
+                  .withEmptyHeaders()
+                  .setRequestBody(
+                      ByteString.copyFromUtf8(
+                          String.format(
+                              HTTP_PARAMETERS, URLEncoder.encode(cmd, StandardCharsets.UTF_8))))
+                  .build(),
+              networkService);
+    } catch (IOException | AssertionError e) {
+      logger.atWarning().withCause(e).log("Request to target %s failed", networkService);
+    }
+    if (response == null || response.bodyString().isEmpty()) {
+      return false;
+    }
+
+    return payload.checkIfExecuted(response.bodyString().get());

```

Comment:
I'm sorry to reopen this as I had already marked it as LGTM but I just noticed that here the sleep must be put back inside an if that guards the instruction and checks if the payload is using OOB (something like `payload.getPayloadAttributes().getUsesCallbackServer()` should work). Tsunami generates reflective payloads only when callback server is not available so here we would miss the detection if it was configured. Please implement the sleep using  `sleepUninterruptibly` and the relative annotations.